### ISSUE TITLE
generalize gradient descent step

### ISSUE THREAD

**Issue by jpreiss** [jpreiss/mlpfile#6]
Currently it's fixed to use the squared error loss. I prototyped:

- using an arbitrary `std::function`, which pybind11 can wrap to call an arbitrary Python function. But it causes slowdown.
- an enum for loss functions. This is limiting, but fast and user-friendly if your loss is implemented.

Still trying to figure out the best blend of generality, speed, and ease of use.

**Comment by jpreiss** [comment]
added in #8 - found that the `std::function` overhead is mainly when going thru pybind11's bridge. But if I make my own set of common loss functions in C++ and bind them (approx same amt of work as the enum) it's fast and general at the same time.
